[DOC] Nebari Troubleshooting page (#97)
nebari-dev/nebari-docs: docs/docs/troubleshooting.md

3. Run the following command: `aws eks --region <region-code> update-kubeconfig --name <project-name>-<namespace>`

After completing these steps according to your cloud provider, `kubectl` should be able to access your Kubernetes cluster.

### Debug your Kubernetes Cluster

If you need more information about your Kubernetes cluster, [`k9s`](https://k9scli.io/) is a terminal-based UI that is extremely useful for debugging. It simplifies navigating, observing, and managing your applications in Kubernetes. `k9s` continuously monitors Kubernetes clusters for changes and provides shortcut commands to interact with the observed resources. It's a fast way to review and resolve day-to-day issues in Kubernetes, a huge improvement to the general workflow, and a best-to-have tool for debugging your Kubernetes cluster sessions.

[Installation instructions for macOS, Windows, and Linux](https://github.com/derailed/k9s) are available.

By default, `k9s` starts with the standard directory that's set as the context (in this case Minikube). To view all the current process press 0:

> ![Image of the k9s terminal UI](../static/images/k9s_UI.png)

Reviewer: ```suggestion
![Image of the k9s terminal UI](../static/img/k9s_UI.png)
```

Author: Thank you!

# Troubleshooting

Invariably you will encounter behavior that does not match your expectations. This guide is meant to explain that behavior, help you diagnose it, and if possible resolve it.

[_Behavior_](#behavior)

 - DNS domain=`your_nebari_domain` record does not exist
 - User instances on AWS occasionally die ~30 minutes after spinning up a large dask cluster
 - Why is the `NEBARI_KUBECONFIG` file in `/tmp`?
 - Required pins for dask environments

[_Errors_](#errors)
 
 - Conda-Store environment fails to build

[_How do I...?_](#how-do-i)

 - Get kubernetes context
 - Debug your kubernetes cluster
 - Deploy an arbitrary pod
 - Upgrade the instance size the `general` node group
 - Use a DNS provider other than CloudFlare
 - Add system packages to the user's (jupyterlab) image
 - Provide individual users with unique environments and cloud instance types

> 

Author: Done.

> ## Behavior

Author: Two of the items here really fit more under "Errors" anyway, making "Configuration" more accurate for the ones that remain.

> ### DNS domain=`your_nebari_domain` record does not exist

Author: Done.

> During your initial Nebari deployment, at the end of the `04-kubernetes-ingress` stage, you may receive an output message stating that the DNS record for `your_nebari_domain` "appears not to exist, has recently been updated, or has yet to fully propagate."

Reviewer: ```suggestion
During your initial Nebari deployment, you may receive an output message stating that the DNS record for `your_nebari_domain` "appears not to exist, has recently been updated, or has yet to fully propagate."
```

Reviewer 2: yup, looks great

At this point, feel free to cancel the deployment and rerun the same deployment command again in an hour or two. Although not guaranteed, it's extremely likely that the DNS will resolve correctly after this prolonged wait period.

### User instances on AWS occasionally die ~30 minutes after spinning up a large dask cluster

AWS uses Elastic Kubernetes Service for hosting the Kubernetes cluster. [It requires the use of at least two availability zones](https://docs.aws.amazon.com/eks/latest/userguide/infrastructure-security.html). The Nebari cluster has an autoscaler with a default service that automatically balances the number of EC2 instances between the two availability zones. When large dask clusters are initialized and destroyed, the autoscaler attempts to reschedule a user pod. This reschedule operation occurs in the other availability zone. Unfortunately, Kubernetes doesn't successfully transfer the active pod to the other zone, and the pod dies.

> To stop this from happening, the autoscaler service "AZRebalance" needs to be manually suspended. You can [follow the steps in the AWS documentation to suspend the AZRebalance service](https://docs.aws.amazon.com/autoscaling/ec2/userguide/as-suspend-resume-processes.html).

Author: From the original tickets/PR it appears to be permanently suspended. However I did ping @costrouc asking for feedback a couple weeks ago and haven't heard back yet ( https://github.com/nebari-dev/nebari-docs/issues/83#issuecomment-1205094129 ). @viniciusdc do you have any additional context here?

### Why is the `NEBARI_KUBECONFIG` file in `/tmp`?

> Nebari regenerates this file on every run. Yes, it will be trashed by the operating system's tmpdir cleanup, but it will be repopulated.

Reviewer 2: that's correct, though I would prefer to present more details on how it can be generated as a reference:
```
Since the `KUBECONFIG_FILE` is stored in `/temp`, the system will remove that file during its clean-up process. If you want to re-generate the file again, you can run the `qhub deploy` command again as Nebari will update/create a `KUBECONFIG_FILE` in for you for each deployment.
```

> Nebari regenerates this file on every run. Yes, it will be trashed by the operating system's tmpdir cleanup, but it will be repopulated.

Author: Done.

If you need more information about your Kubernetes cluster, [`k9s`](https://k9scli.io/) is a terminal-based UI that is extremely useful for debugging. It simplifies navigating, observing, and managing your applications in Kubernetes. `k9s` continuously monitors Kubernetes clusters for changes and provides shortcut commands to interact with the observed resources. It's a fast way to review and resolve day-to-day issues in Kubernetes, a huge improvement to the general workflow, and a best-to-have tool for debugging your Kubernetes cluster sessions.

[Installation instructions for macOS, Windows, and Linux](https://github.com/derailed/k9s) are available.

By default, `k9s` starts with the standard directory that's set as the context (in this case Minikube). To view all the current process press 0:

![Image of the k9s terminal UI](../static/img/k9s_UI.png)

> **NOTE**: In some circumstances, you will be confronted with the need to inspect any services launched by your cluster at your `localhost`. For instance, if your cluster has problem with the network traffic tunnel configuration, it may limit or block the user’s access to destination resources over the connection.

> `k9s` port-forward option <kbd>shift</kbd> + <kbd>f</kbd> allows you to access and interact with internal Kubernetes cluster processes from your localhost you can then use this method to investigate issues and adjust your services locally without the need to expose them beforehand.

Reviewer 2: not sure, it was originally something Tania proposed in the original docs

> `k9s` port-forward option <kbd>shift</kbd> + <kbd>f</kbd> allows you to access and interact with internal Kubernetes cluster processes from your localhost you can then use this method to investigate issues and adjust your services locally without the need to expose them beforehand.

Author: Would backticks be preferable?

> `k9s` port-forward option <kbd>shift</kbd> + <kbd>f</kbd> allows you to access and interact with internal Kubernetes cluster processes from your localhost you can then use this method to investigate issues and adjust your services locally without the need to expose them beforehand.

Reviewer: ah, I see the slight difference between kbd and backticks now. It has a slight border than makes it look like a button! Neat. Yeah leave this in.

### Deploy an arbitrary pod

As a user, you can add extensions as follows:

```sh
extensions:
  - name: echo-test
    image: inanimate/echo-server:latest
    urlslug: echo
    private: true

> ```

Author: Good question. This comes from this QHub Admin FAQ: https://docs.qhub.dev/en/stable/source/admin_guide/faq.html#can-a-user-deploy-an-arbitrary-pod . I'm assuming somewhere in the `qhub-config`, @costrouc or @viniciusdc can you confirm?

> ```

Author: According to Vini and Eskild: we don't even check for extensions anymore! Better way to do this would be to do `kubectl apply -f <pod manifest>`. Will add this in place of the current instructions, thank you Vini and Eskild.

| Cloud Provider | `general` node upgrade possible? |
| :------------- | :------------------------------- |
| AWS            | No (Danger!)                     |
| Azure          | No (Danger!)                     |
| Digital Ocean  | No                               |
| GCP            | Yes                              |

If modifying the resource allocation for the `general` node in-place is absolutely necessary, try increasing the maximum number of nodes for the `general` node group. This will mean two nodes (reserved for the `general` node group) will always be running, ultimately increasing the operating cost of the cluster.

Alternatively, you can backup your cluster, destroy it, specify the new instance size in your `qhub-config.yaml`, and redeploy.

### Use a DNS provider other than CloudFlare

CloudFlare is one of the most commonly used DNS providers for Nebari, so to some it may seem as if it is the *only* DNS provider Nebari supports. This is NOT the case! Please check out our "How to" section, paying particular attention to the domain registry section.

> ### Add system packages to a user's (jupyterlab) image

Reviewer: Why is `jupyterlab` image in parenthesis? Also, its `JupyterLab`

Author: Sorry, taken from this Discussion https://github.com/Quansight/qhub/discussions/712

> In some cases, you may wish to customize the default user's JupyterLab image, including installing some system packages via apt (or other OS package manager) or adding some JupyterLab extensions.

Reviewer: ```suggestion
In some cases, you may wish to customize the default user's JupyterLab image, such as installing system packages via apt (or other OS package manager) or adding JupyterLab extensions.
```

Nebari uses its own registered docker images for `jupyterhub`, `dask`, and `jupyterlab` services by default, but this can be changed by:

 1. Building your own docker images, and
 2. Including the DockerHub register hash into `qhub-config.yaml`.

> ### Provide individual users with unique environments and cloud instance types

Reviewer: What you have here is basically describing nebari in general. I think what you mean is something like:

>Nebari allows for admins to set up user groups which each have access to specific environments and server instance types. This allows for a fine-tuned management of your cloud resources. For example, you can create a special user group for a team of ML engineers which provides access to GPUs and a pytorch environment. This will prevent inexperienced users from accidentally consuming expensive resources. 

and so on...

Author: True, but this came out of a Discussion where that seemed to be non-obvious https://github.com/Quansight/qhub/discussions/675 .

Reviewer: He's running into some sort of error, unrelated to this concept. I guess we can just leave this in though.

> It is common for the user base of a common Nebari deployment to need vastly differing environments and, therefore, cloud instance types for their environments. Users can choose both instance types as well as environments at server launch time, provided you have performed some setup ahead of time. 

Reviewer: ```suggestion
It is common for the user base of a Nebari deployment to need vastly different environments and, therefore, cloud instance types for their environments. Users can choose both instance types as well as environments at server launch time, provided you have performed some setup ahead of time. 
```
